From cd69a2a516e6503c1a1c7e097ee90d255ec3d6b7 Mon Sep 17 00:00:00 2001
From: Nicolas Brassard <nitriques@users.noreply.github.com>
Date: Tue, 9 May 2017 11:29:30 -0400
Subject: [PATCH] Prevent XSS with section's name and nav group

This commit adds sanitization of the section's name and naviguation
group, which permitted authenticated XSS.

Reported by Pradeep Kumar <pradeepch99@gmail.com>

cc @michael-e @brendo
---
 symphony/content/content.blueprintssections.php   | 12 ++++++------
 symphony/content/content.publish.php              | 12 ++++++------
 symphony/lib/toolkit/class.administrationpage.php |  4 ++--
 3 files changed, 14 insertions(+), 14 deletions(-)

diff --git a/symphony/content/content.blueprintssections.php b/symphony/content/content.blueprintssections.php
index 0521ee1a1..115f23ee1 100644
--- a/symphony/content/content.blueprintssections.php
+++ b/symphony/content/content.blueprintssections.php
@@ -215,7 +215,7 @@ public function __viewNew()
 
         $sections = SectionManager::fetch(null, 'ASC', 'sortorder');
         $label = Widget::Label(__('Navigation Group'));
-        $label->appendChild(Widget::Input('meta[navigation_group]', $meta['navigation_group']));
+        $label->appendChild(Widget::Input('meta[navigation_group]', (isset($meta['navigation_group']) ? General::sanitize($meta['navigation_group']) : null)));
 
         if (isset($this->_errors['navigation_group'])) {
             $navgroupdiv->appendChild(Widget::Error($label, $this->_errors['navigation_group']));
@@ -232,7 +232,7 @@ public function __viewNew()
                     continue;
                 }
 
-                $ul->appendChild(new XMLElement('li', $s->get('navigation_group')));
+                $ul->appendChild(new XMLElement('li', General::sanitize($s->get('navigation_group'))));
                 $groups[] = $s->get('navigation_group');
             }
 
@@ -381,8 +381,8 @@ public function __viewEdit()
         }
 
         $this->setPageType('form');
-        $this->setTitle(__('%1$s &ndash; %2$s &ndash; %3$s', array($meta['name'], __('Sections'), __('Symphony'))));
-        $this->appendSubheading($meta['name'],
+        $this->setTitle(__('%1$s &ndash; %2$s &ndash; %3$s', array(General::sanitize($meta['name']), __('Sections'), __('Symphony'))));
+        $this->appendSubheading(General::sanitize($meta['name']),
             Widget::Anchor(__('View Entries'), SYMPHONY_URL . '/publish/' . $section->get('handle'), __('View Section Entries'), 'button')
         );
         $this->insertBreadcrumbs(array(
@@ -425,7 +425,7 @@ public function __viewEdit()
 
         $sections = SectionManager::fetch(null, 'ASC', 'sortorder');
         $label = Widget::Label(__('Navigation Group'));
-        $label->appendChild(Widget::Input('meta[navigation_group]', $meta['navigation_group']));
+        $label->appendChild(Widget::Input('meta[navigation_group]', General::sanitize($meta['navigation_group'])));
 
         if (isset($this->_errors['navigation_group'])) {
             $navgroupdiv->appendChild(Widget::Error($label, $this->_errors['navigation_group']));
@@ -442,7 +442,7 @@ public function __viewEdit()
                     continue;
                 }
 
-                $ul->appendChild(new XMLElement('li', $s->get('navigation_group')));
+                $ul->appendChild(new XMLElement('li', General::sanitize($s->get('navigation_group'))));
                 $groups[] = $s->get('navigation_group');
             }
 
diff --git a/symphony/content/content.publish.php b/symphony/content/content.publish.php
index 3b8930182..c0160bf9c 100644
--- a/symphony/content/content.publish.php
+++ b/symphony/content/content.publish.php
@@ -379,7 +379,7 @@ public function __viewIndex()
         $section = SectionManager::fetch($section_id);
 
         $this->setPageType('table');
-        $this->setTitle(__('%1$s &ndash; %2$s', array($section->get('name'), __('Symphony'))));
+        $this->setTitle(__('%1$s &ndash; %2$s', array(General::sanitize($section->get('name')), __('Symphony'))));
 
         $filters = array();
         $filter_querystring = $prepopulate_querystring = $where = $joins = null;
@@ -469,7 +469,7 @@ public function __viewIndex()
             array_unshift($subheading_buttons, Widget::Anchor(__('Edit Section'), SYMPHONY_URL . '/blueprints/sections/edit/' . $section_id . '/', __('Edit Section Configuration'), 'button'));
         }
 
-        $this->appendSubheading($section->get('name'), $subheading_buttons);
+        $this->appendSubheading(General::sanitize($section->get('name')), $subheading_buttons);
 
         /**
          * Allows adjustments to be made to the SQL where and joins statements
@@ -960,7 +960,7 @@ public function __viewNew()
         $section = SectionManager::fetch($section_id);
 
         $this->setPageType('form');
-        $this->setTitle(__('%1$s &ndash; %2$s', array($section->get('name'), __('Symphony'))));
+        $this->setTitle(__('%1$s &ndash; %2$s', array(General::sanitize($section->get('name')), __('Symphony'))));
 
         // Ensure errored entries still maintain any prepopulated values [#2211]
         $this->Form->setAttribute('action', $this->Form->getAttribute('action') . $this->getPrepopulateString());
@@ -986,7 +986,7 @@ public function __viewNew()
 
         // Build filtered breadcrumb [#1378}
         $this->insertBreadcrumbs(array(
-            Widget::Anchor($section->get('name'), SYMPHONY_URL . '/publish/' . $this->_context['section_handle'] . '/' . $this->getFilterString()),
+            Widget::Anchor(General::sanitize($section->get('name')), SYMPHONY_URL . '/publish/' . $this->_context['section_handle'] . '/' . $this->getFilterString()),
         ));
 
         $this->Form->appendChild(Widget::Input('MAX_FILE_SIZE', Symphony::Configuration()->get('max_upload_size', 'admin'), 'hidden'));
@@ -1297,7 +1297,7 @@ public function __viewEdit()
 
         $this->setPageType('form');
         $this->Form->setAttribute('enctype', 'multipart/form-data');
-        $this->setTitle(__('%1$s &ndash; %2$s &ndash; %3$s', array($title, $section->get('name'), __('Symphony'))));
+        $this->setTitle(__('%1$s &ndash; %2$s &ndash; %3$s', array($title, General::sanitize($section->get('name')), __('Symphony'))));
 
         $sidebar_fields = $section->fetchFields(null, 'sidebar');
         $main_fields = $section->fetchFields(null, 'main');
@@ -1316,7 +1316,7 @@ public function __viewEdit()
         }
 
         $this->insertBreadcrumbs(array(
-            Widget::Anchor($section->get('name'), SYMPHONY_URL . (isset($filter_link) ? $filter_link : $base)),
+            Widget::Anchor(General::sanitize($section->get('name')), SYMPHONY_URL . (isset($filter_link) ? $filter_link : $base)),
         ));
 
         $this->Form->appendChild(Widget::Input('MAX_FILE_SIZE', Symphony::Configuration()->get('max_upload_size', 'admin'), 'hidden'));
diff --git a/symphony/lib/toolkit/class.administrationpage.php b/symphony/lib/toolkit/class.administrationpage.php
index dd82787df..bd9cafa30 100644
--- a/symphony/lib/toolkit/class.administrationpage.php
+++ b/symphony/lib/toolkit/class.administrationpage.php
@@ -807,7 +807,7 @@ public function appendNavigation()
             }
 
             if ($this->doesAuthorHaveAccess($n['limit'])) {
-                $xGroup = new XMLElement('li', $n['name'], array('role' => 'presentation'));
+                $xGroup = new XMLElement('li', General::sanitize($n['name']), array('role' => 'presentation'));
 
                 if (isset($n['class']) && trim($n['name']) !== '') {
                     $xGroup->setAttribute('class', $n['class']);
@@ -826,7 +826,7 @@ public function appendNavigation()
                         if ($this->doesAuthorHaveAccess($c['limit'])) {
                             $xChild = new XMLElement('li');
                             $xChild->setAttribute('role', 'menuitem');
-                            $linkChild = Widget::Anchor($c['name'], SYMPHONY_URL . $c['link']);
+                            $linkChild = Widget::Anchor(General::sanitize($c['name']), SYMPHONY_URL . $c['link']);
                             if (isset($c['target'])) {
                                 $linkChild->setAttribute('target', $c['target']);
                             }
